x/evm/keeper/grpc_query.go

```diff
@@ -73,66 +76,73 @@ func (q Querier) BurnerInfo(c context.Context, req *types.BurnerInfoRequest) (*t
 	return nil, status.Error(codes.NotFound, "unknown address")
 }
 
-func getExecuteDataAndSigs(ctx sdk.Context, s types.Signer, commandBatch types.CommandBatch) ([]byte, []types.Signature, error) {
-	id := hex.EncodeToString(commandBatch.GetID())
-	sig, sigStatus := s.GetSig(ctx, id)
-	if sigStatus != tss.SigStatus_Signed {
-		return nil, nil, fmt.Errorf("could not find a corresponding signature for sig ID %s", id)
+func optimizeSignatureSet(key multisig.Key, operators []types.Operator) []types.Operator {
+	sort.SliceStable(operators, func(i, j int) bool {
+		return operators[i].GetWeight().GT(operators[j].GetWeight())
+	})
+
+	cumWeight := sdk.ZeroUint()
+	for _, o := range operators {
+		if cumWeight.GTE(key.GetMinPassingWeight()) {
+			o.Signature = nil
+		} else if o.Signature != nil {
+			cumWeight = cumWeight.Add(o.Weight)
+		}
 	}
 
-	switch sig := sig.GetSig().(type) {
-	case *tss.Signature_SingleSig_:
-		commandBatchSig, err := getCommandBatchSig(sig.SingleSig.SigKeyPair, commandBatch.GetSigHash())
-		if err != nil {
-			return nil, nil, err
-		}
+	sort.SliceStable(operators, func(i, j int) bool {
+		return bytes.Compare(operators[i].Address.Bytes(), operators[j].Address.Bytes()) < 0
+	})
 
-		executeData, err := types.CreateExecuteDataSinglesig(commandBatch.GetData(), commandBatchSig)
-		if err != nil {
-			return nil, nil, fmt.Errorf("could not create transaction data: %s", err)
-		}
+	return operators
+}
 
-		return executeData, []types.Signature{commandBatchSig}, nil
-	case *tss.Signature_MultiSig_:
-		sigKeyPairs := types.SigKeyPairs(sig.MultiSig.SigKeyPairs)
-		sort.Stable(sigKeyPairs)
+func getProof(ctx sdk.Context, commandBatch types.CommandBatch, s types.MultisigKeeper) (types.AuthData, error) {
```

Comment:
```suggestion
func getProof(ctx sdk.Context, commandBatch types.CommandBatch, s types.MultisigKeeper) ([]types.Address, []sdk.Uint, sdk.Uint, [][]byte) {
	multiSig := commandBatch.GetSignature().(*multisigtypes.MultiSig)
	key := funcs.MustOk(s.GetKey(ctx, multiSig.KeyID))
	addressWeights, threshold := types.ParseMultisigKey(key)

	addressesByWeightDesc := maps.Keys(addressWeights)
	sort.SliceStable(addressesByWeightDesc, func(i, j int) bool {
		return addressWeights[addressesByWeightDesc[i]].GT(addressWeights[addressesByWeightDesc[j]])
	})

	addresses := make([]types.Address, 0, len(addressesByWeightDesc))
	sumWeight := sdk.ZeroUint()
	for _, addr := range addressesByWeightDesc {
		if sumWeight.GTE(threshold) {
			break
		}

		sumWeight = sumWeight.Add(addressWeights[addr])
		addresses = append(addresses, types.Address(common.HexToAddress(addr)))
	}

	sort.SliceStable(addresses, func(i, j int) bool {
		return bytes.Compare(addresses[i].Bytes(), addresses[j].Bytes()) < 0
	})

	addressSigs := make(map[string][]byte, len(multiSig.Sigs))
	for p, sig := range commandBatch.GetSignature().(*multisigtypes.MultiSig).Sigs {
		address := crypto.PubkeyToAddress(funcs.MustOk(key.GetPubKey(funcs.Must(sdk.ValAddressFromBech32(p)))).GetECDSAPubKey())
		addressSigs[address.Hex()] = funcs.Must(types.NewSignature(sig)).ToHomesteadSig()
	}

	return addresses,
		slices.Map(addresses, func(address types.Address) sdk.Uint { return addressWeights[address.Hex()] }),
		threshold,
		slices.Map(addresses, func(address types.Address) []byte { return addressSigs[address.Hex()] })
}
```

Reply:
so that you don't have to define types.Auth and types.Operator. I just find both to be hard to reason about.